File: backend/src/application/agent/OptimizationSuggestionGenerator.ts
import { ListingAnalysis } from "./ListingAnalysis.js";
import { OptimizationDraft } from "./OptimizationDraft.js";
import { OptimizationProvider } from "./OptimizationProvider.js";
import { ProductSnapshot } from "./ProductSnapshot.js";

export class OptimizationSuggestionGenerator {
  constructor(private readonly providers: OptimizationProvider[]) {}

  async generate(snapshot: ProductSnapshot, analysis: ListingAnalysis): Promise<OptimizationDraft> {
    for (const provider of this.providers) {
      const draft = await provider.generate(snapshot, analysis);

      if (draft) {
        return this.normalize(draft);
      }
    }

    throw new Error("No optimization provider produced a suggestion.");
  }

  private normalize(draft: OptimizationDraft): OptimizationDraft {
    return {
      improvedTitle: draft.improvedTitle?.trim() || null,
      improvedDescription: draft.improvedDescription.trim(),
      tags: [...new Set(draft.tags.map((tag) => tag.trim().toLowerCase()).filter(Boolean))].slice(0, 12),
      positioningRecommendations: draft.positioningRecommendations?.trim() || null,
      seoNotes: draft.seoNotes?.trim() || null,
      score: draft.score === null ? null : Math.max(0, Math.min(Number(draft.score), 100)),
      rationale: draft.rationale?.trim() || null
    };
  }
}

File: app/services/agent/optimization_suggestion_generator.rb
module Agent
  class OptimizationSuggestionGenerator
    def initialize(registry = OptimizationProviderRegistry.new)
      @registry = registry
    end

    def generate(step_name:, prompt:, snapshot:, analysis:, prior_responses: {})
      @registry.providers.each do |provider|
        result = provider.generate(
          step_name: step_name,
          prompt: prompt,
          snapshot: snapshot,
          analysis: analysis,
          prior_responses: prior_responses
        )

        next unless result

        if step_name == "final_synthesis"
          next unless result[:draft]
        end

        return result.merge(draft: normalize_draft(result[:draft]))
      end

      raise "No optimization provider produced a response for #{step_name}."
    end

    private

    def normalize_draft(draft)
      return nil unless draft

      tags = Array(draft[:tags]).map { |t| t.to_s.strip.downcase }.select(&:present?).uniq.first(12)

      {
        improved_title:                draft[:improved_title]&.strip.presence,
        improved_description:          draft[:improved_description].to_s.strip,
        tags:                          tags,
        positioning_recommendations:   draft[:positioning_recommendations]&.strip.presence,
        seo_notes:                     draft[:seo_notes]&.strip.presence,
        score:                         draft[:score] ? [[draft[:score].to_f, 0].max, 100].min : nil,
        rationale:                     draft[:rationale]&.strip.presence
      }
    end
  end
end
